File: my-messenger-backend/my.messenger.common/Messaging/Impl/RabbitMQMessageSender.cs
using Microsoft.Extensions.Configuration;
using System;
using System.Threading.Tasks;

namespace my.messenger.common.Messaging.Impl
{
    public class RabbitMQMessageSender : RabbitMQBaseMessage, IMessageSender
    {
        public RabbitMQMessageSender(
           IConfiguration configuration)
           : base(configuration)
        {
        }

        public Task RegisterDestinationAsync(Destination d)
        {
            return Task.Run(() => useRabbitAndDispose((conn, channel) => {
                String exchangeName = getExchangeName(d);
                channel.ExchangeDeclare(exchangeName, "fanout", true, false, null);

                if (d.Type == DestinationType.User)
                {
                    String queueName = getQueueName(d);

                    channel.QueueDeclare(queueName, true, false, false, null);
                    channel.QueueBind(queueName, exchangeName, "", null);
                }
            }));
        }

        public Task RegisterDestinationListenerAsync(Destination d, Destination destinationListener)
        {
            return Task.Run(() => useRabbitAndDispose((conn, channel) => {
                String exchangeName = getExchangeName(d);

                if (destinationListener.Type == DestinationType.User)
                {
                    String queueName = getQueueName(destinationListener);
                    channel.QueueBind(queueName, exchangeName, "", null);
                }
            }));
        }

        public Task RemoveDestinationListenerAsync(Destination d, Destination destinationListener)
        {
            return Task.Run(() => useRabbitAndDispose((conn, channel) => {
                String exchangeName = getExchangeName(d);

                if (destinationListener.Type == DestinationType.User)
                {
                    String queueName = getQueueName(destinationListener);
                    channel.QueueUnbind(queueName, exchangeName, "", null);
                }
            }));
        }

        public Task SendAync(Message msg)
        {
            return Task.Run(() => useRabbitAndDispose((conn, channel) => {
                String exchangeName = getExchangeName(msg.To);

                byte[] messageBodyBytes = JsonUtil.ToBytes(msg);

                var properties = channel.CreateBasicProperties();
                properties.Persistent = true;

                channel.BasicPublish(
                        exchangeName, 
                        "",
                        false,
                        properties,
                        messageBodyBytes);
            }));
        }
    }
}

File: my-messenger-backend-java/src/main/java/mymessenger/backend/messaging/impl/RabbitMQMessageSender.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package mymessenger.backend.messaging.impl;

import com.rabbitmq.client.AMQP;
import mymessenger.backend.encoding.JsonUtil;
import mymessenger.backend.messaging.MessageSender;
import mymessenger.backend.model.messaging.Destination;
import mymessenger.backend.model.messaging.Message;

import mymessenger.backend.messaging.MessageSenderException;
import org.springframework.stereotype.Service;

/**
 *
 * @author guilherme
 */
@Service
public class RabbitMQMessageSender extends RabbitMQBaseMessage implements MessageSender {

    @Override
    public void registerDestination(Destination d) throws MessageSenderException {
        useRabbitAndDispose(rabbitMq -> {
            String exchangeName = getExchangeName(d);
            String queueName = getQueueName(d);
            rabbitMq.channel.exchangeDeclare(exchangeName, "fanout", true);
            /*
                    String queue,
                    boolean durable,
                    boolean exclusive,
                    boolean autoDelete,
                    Map<String,Object> arguments
             */
            rabbitMq.channel.queueDeclare(queueName, true, false, false, null);

            rabbitMq.channel.queueBind(queueName, exchangeName, "");
        });
    }

    @Override
    public void send(Message msg) throws MessageSenderException {
        useRabbitAndDispose(rabbitMq -> {
            String exchangeName = getExchangeName(msg.getTo());

            byte[] messageBodyBytes = JsonUtil.toBytes(msg);

            rabbitMq.channel.basicPublish(exchangeName, "",
                    new AMQP.BasicProperties.Builder()
                            .deliveryMode(2) // persistent
                            .build(),
                    messageBodyBytes);
        });
    }

}
